src/notification/NotificationController.ts

```diff
@@ -0,0 +1,120 @@
+import BaseController, { BaseConfig, BaseState } from '../BaseController';
+
+type action = () => void;
+
+interface viewedNotification {
+  [id: number]: boolean;
+}
+
+export interface Notification{
+  id: number;
+  title: string;
+  description: string;
+  date: string;
+  image?: string;
+  actionText: string;
+  isShown?: boolean;
+}
+
+/**
+ * NotitificationConfig will hold the notifications from JSON file read
+ * from `metamask-extension`
+ */
+export interface NotificationConfig extends BaseConfig{
+  allNotifications: Record<string, unknown>[];
+}
+
+/**
+ * Notification state will hold all the seen and unseen notifications
+ * that are still active
+ */
+export interface NotificationState extends BaseState{
+  notifications: { [id: number]: Notification};
+}
+
+const defaultState = {};
+
+/**
+ * Controller for managing in-app announcement notifications.
+ */
+export class NotificationController extends BaseController<NotificationConfig, NotificationState> {
+
+  private readonly allNotifications: Record<string, unknown>[];
+
+  /**
+   * Creates a NotificationController instance
+   *
+   * @param config - Initial options used to configure this controller
+   * @param state - Initial state to set on this controller
+   */
+  constructor(config: NotificationConfig, state?: NotificationState) {
+    const { allNotifications } = config;
+    super(config, state || defaultState);
+    this.allNotifications = [...allNotifications];
+    this.initialize();
+    this._addNotifications();
+  }
+
+  /**
+   * Compares the notifications in state with the notifications from file
+   * to check if there are any new notifications/announcements
+   * if yes, the new notification will be added to the state with a flag indicating
+   * that the notification is not seen by the user.
+   *
+   *  @param allNotifications
+   */
+  private _addNotifications(): void{
+    const existingNotificationIds: number[] = this.state.notifications ? Object.keys(this.state.notifications).map(Number) : [];
+
+    const newNotifications: Notification[] = this.allNotifications
+      .filter((newNotification) => !existingNotificationIds
+      .some((existingId) => (existingId === newNotification.id)))
+      .map((notification) => {
+        if (!((notification.actionText as string).length > 0 && notification.hasOwnProperty('actionFunction'))) {
```

Comment:
This logic is correct, but I think there is an elegant way to ensure this condition is met just by using the type definition. Could be to have an `action` property which is an object, which is optional, but which itself has two properties (function and text) that are both required. You might have to read into some typescript documentation